### ISSUE TITLE
Reproduction Imagenet-1K code - Problems

### ISSUE THREAD

**Issue by alexcbb** [rbalestr-lab/lejepa#25]
Hello, 

I've been trying to reproduce the pre-training of the model on the ImageNet-1K dataset in order to have a checkpoint to experiment with.
My first experiments where a bit unsuccessful with the following code I've setup with Pytorch-Lightning:

```
import torch, torch.nn as nn, torch.nn.functional as F
from torch.utils.data import DataLoader
from torchvision.transforms import v2
import timm, wandb, hydra
from omegaconf import DictConfig
from datasets import load_dataset
from torch.optim.lr_scheduler import LinearLR, CosineAnnealingLR, SequentialLR
from torchvision.ops import MLP
import lightning as L
from lightning.pytorch.loggers import WandbLogger
from lightning.pytorch.callbacks import LearningRateMonitor, ModelCheckpoint
from custom_transforms import CustomDataAugmentation, collate_fn
from torchvision import datasets

class SIGReg(torch.nn.Module):
    def __init__(self, window=(-5, 5), knots=17, num_slices=1024):
        super().__init__()
        self.num_slices = num_slices
        t = torch.linspace(window[0], window[1], knots, dtype=torch.float32)
        dt = 3 / (knots - 1)
        weights = torch.full((knots,), 2 * dt, dtype=torch.float32)
        weights[[0, -1]] = dt
        window = torch.exp(-t.square() / 2.0)
        self.register_buffer("t", t)
        self.register_buffer("phi", window)
        self.register_buffer("weights", weights * window)

    def forward(self, proj):
        A = torch.randn(proj.size(-1), self.num_slices, device="cuda")
        A = A.div_(A.norm(p=2, dim=0))
        x_t = (proj @ A).unsqueeze(-1) * self.t
        err = (x_t.cos().mean(-3) - self.phi).square() + x_t.sin().mean(-3).square()
        statistic = (err @ self.weights) * proj.size(-2)
        return statistic.mean()

class ViTEncoder(nn.Module):
    def __init__(self, cfg, proj_dim=128):
        super().__init__()
        self.backbone = timm.create_model(
            cfg.encoder,
            pretrained=False,
            num_classes=512,
            drop_path_rate=0.1,
            img_size=cfg.img_size,
            dynamic_img_size=True
        )
        self.proj = MLP(512, [2048, 2048, proj_dim], norm_layer=nn.BatchNorm1d)

    def forward(self, x):
        if len(x.shape) == 5:
            N, V = x.shape[:2]
        else:
            N = x.shape[0]
            x = x.unsqueeze(1)
            V = 1

        # Follow multi-crop processing from DINO:        
        x = x.flatten(0, 1)
        x = [x]
        idx_crops = torch.cumsum(torch.unique_consecutive(
            torch.tensor([inp.shape[-1] for inp in x]),
            return_counts=True,
        )[1], 0)
        start_idx, output = 0, torch.empty(0).to(x[0].device)
        for end_idx in idx_crops:
            _out = self.backbone(torch.cat(x[start_idx: end_idx]))
            emb = torch.cat((output, _out))
            start_idx = end_idx
        return emb, self.proj(emb)

class ResNet34Encoder(nn.Module):
    def __init__(self, proj_dim: int = 128) -> None:
        super().__init__()
        self.backbone = timm.create_model(
            "resnet34",
            pretrained=False,
            num_classes=512,
        )
        self.proj = MLP(512, [2048, 2048, proj_dim], norm_layer=nn.BatchNorm1d)

    def forward(self, x):
        n, v = x.shape[:2]
        emb = self.backbone(x.flatten(0, 1))
        out = self.proj(emb).reshape(n, v, -1).transpose(0, 1)
        return emb, out
    
class LeJEPAAugmentation:
    def __init__(self, global_crop_size=224, local_crop_size=96, global_crops_scale=(0.3, 1.0), local_crops_scale=(0.05, 0.3), local_crops_number=6):
        self.local_crops_number = local_crops_number
        
        # Base augmentations (color jitter, etc.)
        flip_and_color_jitter = v2.Compose([
            v2.RandomHorizontalFlip(p=0.5),
            v2.RandomApply([v2.ColorJitter(0.8, 0.8, 0.8, 0.2)], p=0.8),
            v2.RandomGrayscale(p=0.2),
        ])
        
        # Normalize
        normalize = v2.Compose([
            v2.ToImage(),
            v2.ToDtype(torch.float32, scale=True),
            v2.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
        ])

        # Global Views (224x224, Large Scale)
        self.global_transfo = v2.Compose([
            v2.RandomResizedCrop(global_crop_size, scale=global_crops_scale, interpolation=3),
            flip_and_color_jitter,
            v2.RandomApply([v2.GaussianBlur(kernel_size=7, sigma=(0.1, 2.0))], p=1.0), # Stronger blur for global
            normalize,
        ])

        # Local Views (96x96, Small Scale)
        self.local_transfo = v2.Compose([
            v2.RandomResizedCrop(local_crop_size, scale=local_crops_scale, interpolation=3),
            flip_and_color_jitter,
            v2.RandomApply([v2.GaussianBlur(kernel_size=7, sigma=(0.1, 2.0))], p=0.5),
            normalize,
        ])

    def __call__(self, img):
        crops = []
        crops.append(self.global_transfo(img))
        crops.append(self.global_transfo(img))
        for _ in range(self.local_crops_number):
            crops.append(self.local_transfo(img))
        return crops

class HFDataset(torch.utils.data.Dataset):
    def __init__(self, dataset_name, split):
        self.ds = load_dataset(dataset_name, split=split)
        self.is_train = split == "train"
        if self.is_train:
            self.transform = LeJEPAAugmentation()
        else:
            self.transform = v2.Compose(
                [
                    v2.Resize(256),
                    v2.CenterCrop(224),
                    v2.ToImage(),
                    v2.ToDtype(torch.float32, scale=True),
                    v2.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
                ]
            )

    def __getitem__(self, i):
        item = self.ds[i]
        img = item["image"].convert("RGB")
        return self.transform(img), item["label"]

    def __len__(self):
        return len(self.ds)

def multi_crop_collate(batch):
    labels = torch.tensor([b[1] for b in batch])
        
    global_crops = [] 
    local_crops = [] 
    
    for item in batch:
        crops = item[0]
        global_crops.extend(crops[:2]) 
        local_crops.extend(crops[2:])
        
    global_crops = torch.stack(global_crops)
    local_crops = torch.stack(local_crops)
    return (global_crops, local_crops), labels

class LeJEPAModule(L.LightningModule):
    def __init__(self, cfg: DictConfig):
        super().__init__()
        self.save_hyperparameters()
        self.automatic_optimization = False
        self.cfg = cfg

        if self.cfg.encoder == "resnet34":
            self.net = ResNet34Encoder(proj_dim=cfg.proj_dim)
        else:
            self.net = ViTEncoder(cfg, proj_dim=cfg.proj_dim)
        self.probe = nn.Sequential(nn.LayerNorm(512), nn.Linear(512, 1000))

        self.sigreg = SIGReg()
    
    def training_step(self, batch, batch_idx):
        opt = self.optimizers()
        scheduler = self.lr_schedulers()
        opt.zero_grad()

        (global_crops, local_crops), y = batch
        batch_size = y.size(0)
        global_crops = global_crops.to(self.device, non_blocking=True)
        local_crops = local_crops.to(self.device, non_blocking=True)
        y = y.to(self.device, non_blocking=True)

        if global_crops.shape[0] != batch_size:
            global_crops = global_crops.view(batch_size, -1, *global_crops.shape[1:])
            local_crops = local_crops.view(batch_size, -1, *local_crops.shape[1:])

        global_emb, global_proj = self.net(global_crops)
        _, local_proj = self.net(local_crops)

        # --- Loss Calculation ---
        batch_size = y.size(0)
        global_proj_reshaped = global_proj.view(2, batch_size, -1)
        global_centroid = global_proj_reshaped.mean(dim=0) # [B, D]
        
        n_local_crops = local_proj.size(0) // batch_size
        global_target = global_centroid.repeat(n_local_crops, 1) # [N_local * B, D]

        pred_loss = F.mse_loss(local_proj, global_target)
        global_consistency = F.mse_loss(global_proj_reshaped[0], global_proj_reshaped[1])
        total_inv_loss = pred_loss + global_consistency

        all_projs = torch.cat([global_proj, local_proj], dim=0)
        sigreg_loss = self.sigreg(all_projs)
        
        lejepa_loss = sigreg_loss * self.cfg.lamb + total_inv_loss * (1 - self.cfg.lamb)
        
        y_rep = y.repeat(2)
        yhat = self.probe(global_emb.detach())
        probe_loss = F.cross_entropy(yhat, y_rep)
        
        loss = lejepa_loss + probe_loss
        self.manual_backward(loss)
        self.clip_gradients(opt, gradient_clip_val=3.0, gradient_clip_algorithm="norm")
        opt.step()
        scheduler.step()

        self.log_dict(
            {
                "train/probe_loss": probe_loss.item(),
                "train/lejepa_loss": lejepa_loss.item(),
                "train/sigreg_loss": sigreg_loss.item(),
                "train/inv_loss": total_inv_loss.item(),
            },
            prog_bar=True,
            on_step=True,
            on_epoch=False,
        )
        return loss
    
    def validation_step(self, batch, batch_idx):
        x, y = batch
        x = x.to(self.device, non_blocking=True)
        y = y.to(self.device, non_blocking=True)
        emb, _ = self.net(x)
        acc = (self.probe(emb).argmax(1) == y).float().mean()
        self.log("val/acc", acc, prog_bar=True)
        return acc

    def configure_optimizers(self):
        param_groups = [
            {"params": self.net.parameters(), "lr": self.cfg.lr, "weight_decay": 5e-2},
            {"params": self.probe.parameters(), "lr": 1e-3, "weight_decay":  1e-6}
        ]
        opt = torch.optim.AdamW(param_groups)
        
        scheduler = SequentialLR(opt, schedulers=[
            LinearLR(opt, start_factor=1e-4, total_iters=self.cfg.warmup_steps),
            CosineAnnealingLR(opt, T_max=self.cfg.total_steps - self.cfg.warmup_steps, eta_min=1e-6)
        ], milestones=[self.cfg.warmup_steps])
        return [opt], [{"scheduler": scheduler, "interval": "step"}]

@hydra.main(config_path="configs", config_name="train_jepa")
def main(cfg: DictConfig):
    torch.manual_seed(0)

    if cfg.dataset == "ILSVRC/imagenet-1k":
        train_ds = HFDataset(cfg.dataset, split="train")
        val_ds = HFDataset(cfg.dataset,split="validation")
        train = DataLoader(
            train_ds, 
            batch_size=cfg.bs, 
            shuffle=True, 
            drop_last=True, 
            num_workers=cfg.num_workers,
            collate_fn=multi_crop_collate
        )
        val = DataLoader(val_ds, batch_size=256, num_workers=cfg.num_workers)
    else:
        raise NotImplementedError(f"Dataset {cfg.dataset} not implemented.")

    print(f"Training on {len(train.dataset)} samples, validating on {len(val.dataset)} samples.")
    model_name = f"LeJEPA_{cfg.encoder}_{cfg.dataset}_lamb{cfg.lamb}_lr{cfg.lr}_bs{cfg.bs}"
    lr_monitor = LearningRateMonitor(logging_interval='step')
    checkpoint_callback = ModelCheckpoint(
        save_top_k=3,
        monitor="lejepa_loss",
        mode="min",
        dirpath=f"./checkpoints/",
        filename=f"{model_name}_"+"{epoch:02d}-{lejepa_loss:.2f}",
        save_weights_only=True, 
        every_n_train_steps=cfg.checkpoint_every_n_steps,
    )
    wandb_logger = WandbLogger(project="LeJEPA", offline=cfg.offline, name=model_name)
    model = LeJEPAModule(cfg)
    print(f"Model has {sum(p.numel() for p in model.parameters() if p.requires_grad)/1e6:.2f}M trainable parameters.")
    trainer = L.Trainer(
        logger=wandb_logger,
        callbacks=[lr_monitor, checkpoint_callback],
        max_epochs=cfg.total_epochs,
        val_check_interval=cfg.val_every_n_steps,
        log_every_n_steps=cfg.log_every_n_steps,
        check_val_every_n_epoch=None,
        precision="16-mixed"
    )
    trainer.fit(model, train, val)

if __name__ == "__main__":
    main()
```

Here is my set of hyperparameters:
```
# model
encoder: vit_small_patch16_224
num_classes: 1000
img_size: 224
lamb: 0.02
proj_dim: 64

# training
lr: 5e-4
bs: 256
epochs: 100
checkpoint_every_n_steps: 50000
total_steps: 400000
total_epochs: 100
warmup_steps: 10000
val_every_n_steps: 50000
log_every_n_steps: 500
offline: true

# data
dataset: ILSVRC/imagenet-1k

# misc
num_workers: 8
```

Do someone might have any clue on what might be wrong on my implementation ? Do you have maybe a proper code for ImageNet1K with the proper set of hyperparameters ? Thank you in advance !

**Comment by alexcbb** [comment]
> Hello! Thanks for your interest! I think a good starting point is to reuse the minimal code but only swap the dataset to inet1k one (is that what you did? I didn't run a diff). Could you share your training loss values? it is possible that you want to
> 
>     * increase the number of views (we used 8 to follow DINO for inet1k)
> 
>     * play with lambda a little bit (especially if you change the number of views)
> 
> 
> Let me know!

Yes I've used the minimal example but made some modifications to handle the multi-view like in DINO (2 global views and 6 local views). I've also setup the loss as MSE between global and local, local and local, and global and global. The thing is that the model seems to diverge after few thousands of steps. I've played a bit with the learning rate but it does not seem to avoid the issue. Do you have maybe some specific hyperparameters you used for training ? Btw I'm using a ViT-small

**Comment by RandallBalestriero** [comment]
I see, for that case the prediction loss we used was the one of the paper, i.e., you try to match all the views (global and locals) to the global ones only (not sure if that's the same as your case). Then is it the loss that diverges or only sigreg/prediction?
- definitely tune down learning rate or add gradient clipping to norm=1 if needed
- play with lambda if the issue is one loss vs the other
- for the online probe, is training stable? maybe if it is the linear probe diverging you can add a batchnorm or layer norm to it!
Feel free to share some loss curves to help you more!

**Comment by RandallBalestriero** [comment]
Closing for now, if you have updates please ping again, thank you @alexcbb !
